runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java

```diff
@@ -31,17 +34,47 @@ public SequencerView(CorfuRuntime runtime) {
      * @return The first token retrieved.
      */
     public TokenResponse nextToken(Set<UUID> streamIDs, int numTokens) {
-        return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
-                .nextToken(streamIDs, numTokens)));
+        if (numTokens == 0) {
+            long tail = checkHints(streamIDs);
+            if (tail != Address.NOT_FOUND) {
+                // the request can be serviced from sequencer hints (omitting sequencer request)
+                long epoch = TransactionalContext.getCurrentContext().hintsEpoch;
+                return new TokenResponse(tail, epoch, null);
+            } else {
+                // the request can't be serviced from the sequencer hints, need to
+                // request the tail from the sequencer
+                return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
+                        .nextToken(streamIDs, numTokens)));
+            }
+        } else {
+            // Not a tail query, by-pass sequencer hints
+            return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
+                    .nextToken(streamIDs, numTokens)));
+        }
     }
 
-
     public TokenResponse nextToken(Set<UUID> streamIDs, int numTokens,
-                                   TxResolutionInfo conflictInfo) {
+                                  TxResolutionInfo conflictInfo) {
         return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
                 .nextToken(streamIDs, numTokens, conflictInfo)));
     }
 
+    /**
+     * check if a stream tail query can be serviced from the sequencer hints.
+     * @param streamIds
+     * @return stream tail or Address.NOT_FOUND if the hints don't exist.
+     */
+    private long checkHints(Set<UUID> streamIds) {
+        AbstractTransactionalContext ctx = TransactionalContext.getCurrentContext();
```

Comment:
Why is the hint located in the transactional context and not the sequencer view itself?

Reply:
The hints are an extension of the transaction time snapshot, a thread in a transaction will only need the stream tails at the time when the transaction time snapshot was acquired.